nixos/modules/config/shells-environment.nix

```diff
@@ -172,6 +172,13 @@ in
          export PATH="$HOME/bin:$PATH"
        '';
 
+    # The set-environment.sh script is exposed here for use by shells in
+    # loading the essential Nix environment variables. It is loaded automatically
+    # by Nixpkgs' fish package, but it can by used by other shells in and
+    # outside of Nixpkgs. This is meant to be a path which can be used in common
+    # among different Nix module systems, for example NixOS and nix-darwin.
+    environment.etc."nix/support/set-environment.sh".source = "${system.build.setEnvironment}";
```

Comment:
Would rename to `nix/support/fish-set-environment.sh`.

Reply:
I have this set in this location because it's identical to the usual setEnvironment script and exposing it this way makes it usable by Nix packages which may not know whether or what module system they have. I'd like for it to potentially be usable by other non-POSIX shells which may have to resort to similar tricks if we're not to maintain several initialization scripts for Nix. (Elvish is the shell I have in mind.)

Maybe I should say I'm not gonna need it, call it `fish-set-environment.sh` and rename it only when it is actually used by another shell. What do you think?

Reply:
I’m not quite sure I understand the reason for exposing this file in the first place. Since it’s exposed via the module system, you have to use NixOS anyway to use it, so why not add it to fish and other shells directly when building their config?

Reply:
Well, `set-environment` is roughly the equivalent of the `profile.d/nix.sh` script on NixOS and in nix-darwin. The idea is that Fish installed via Nix should at a minimum ‘just work’ with the very basics of Nix (setting up the `NIX_…` environment variables and adding the Nix profile directories to the `PATH`.

This PR adds functionality that makes Fish-from-Nixpkgs usable as a login shell OOTB on non-NixOS. Exposing the `setEnvironment` file allows NixOS to have the same level of functionality prior to enabling the NixOS module. But the NixOS module still adds properly NixOS-specific functionality, like reading all the global shell configurations.

For the feature to work the same way across NixOS and non-NixOS, something equivalent to the `profiles.d/nix.sh` file for NixOS has to be exposed prior to user configuration, and as far as I can tell `setEnvironment` is it.

Reply:
Looks like this went in directly under `/etc` [here](https://github.com/NixOS/nixpkgs/pull/45784/files#diff-82a095f8dc31ee61c534cf139301af25R167)